internal/daemon/api_logs.go

```diff
@@ -0,0 +1,275 @@
+	if follow {
+		flushWriter(w) // first flush any logs written above
+
+		// followLogs sends logs to output channel as they arrive. Do this
+		// until request is cancelled.
+		output := make(chan servicelog.Entry)
+		go followLogs(itsByName, req.Context().Done(), output)
+		for entry := range output {
+			err = encoder.Encode(jsonLog(entry))
+			if err != nil {
+				log.Printf("error writing logs: %v", err)
+				return
+			}
+			flushWriter(w)
+		}
+	}
+}
+
+// writeLogs reads logs from the given iterators, merging the log streams and
+// calling writeLog for each log.
+func writeLogs(itsByName map[string]servicelog.Iterator, writeLog writeLogFunc) error {
+	// Stop when we get a log recorded after this time. This prevents
+	// continually trying to catch up if the client is slow at reading them.
+	stop := time.Now().UTC()
+
+	// Service names, ordered alphabetically for consistent output.
+	names := make([]string, 0, len(itsByName))
+	for name := range itsByName {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+
+	parsersByName := make(map[string]*servicelog.Parser, len(itsByName))
+	for name, it := range itsByName {
+		parsersByName[name] = servicelog.NewParser(it, logReaderSize)
+	}
+
+	// Slice of parsers (or nil) holding the next log from each service.
+	parsers := make([]*servicelog.Parser, len(names))
+	for {
+		// Grab next log for iterators that need fetching (parsers[i] nil).
+		for i, name := range names {
+			if parsers[i] == nil {
+				parser := parsersByName[name]
+				if parser.Next() && parser.Entry().Time.Before(stop) {
+					parsers[i] = parser
+				}
+			}
+		}
+
+		// Find log with earliest timestamp. Linear search is okay here as there
+		// will only be a very small number of services.
+		earliest := -1
+		for i, parser := range parsers {
+			if parser == nil {
+				continue
+			}
+			if parser.Err() != nil {
+				return parser.Err()
+			}
+			if earliest < 0 || parser.Entry().Time.Before(parsers[earliest].Entry().Time) {
+				earliest = i
+			}
+		}
+		// We're done when it.Next() returned false for all iterators.
+		if earliest < 0 {
+			break
+		}
+
+		// Write the log
+		err := writeLog(parsers[earliest].Entry())
+		if err != nil {
+			return err
+		}
+
+		// Set iterator to nil so we fetch next log for that service.
+		parsers[earliest] = nil
+	}
+	return nil
+}
+
+type writeLogFunc func(entry servicelog.Entry) error
+
+// followLogs follows ("tails") the log iterators and sends logs to the output
+// channel as they're written. It stops when the done channel is closed, and
+// closes the output channel when everything is finished.
+func followLogs(itsByName map[string]servicelog.Iterator, done <-chan struct{}, output chan<- servicelog.Entry) {
+	// Start one goroutine per service to listen for and output new logs.
+	var wg sync.WaitGroup
+	for _, it := range itsByName {
+		wg.Add(1)
+		go func(it servicelog.Iterator) {
+			defer wg.Done()
+			parser := servicelog.NewParser(it, logReaderSize)
```

Comment:
parser setup should happen outside of followLogs so that writeLogs and followLogs can share the same parser so to not skip data already buffered.

Reply:
Done.